src/ru/practicum/tasktracker/service/InMemoryHistoryManager.java

```diff
@@ -2,26 +2,110 @@
 
 import ru.practicum.tasktracker.model.Task;
 
-import java.util.LinkedList;
+import java.util.Map;
+import java.util.HashMap;
 import java.util.List;
+import java.util.ArrayList;
+
 
 public class InMemoryHistoryManager implements HistoryManager {
-    private final List<Task> history = new LinkedList<>();
-    private static final int MAX_HISTORY_SIZE = 10;
+    private Node first;
+    private Node last;
+    private Map<Integer, Node> nodeMap = new HashMap<>();
 
     @Override
     public void add(Task task) {
         if (task == null) {
             return;
         }
-        if (history.size() == MAX_HISTORY_SIZE) {
-            history.removeFirst();
+        if (nodeMap.containsKey(task.getId())) {
+            remove(task.getId());
+        }
+        linkLast(task);
+    }
+
+    @Override
+    public void remove(int id) {
+        Node node = nodeMap.remove(id);
+        if (node == null) {
+            return;
+        } else {
+            removeNode(node);
         }
```

Comment:
```suggestion
        if (node != null) {
            removeNode(node);
        }
```
можно немного упростить